### PR TITLE
SCP-1654: Hold on to intermediate states when evaluating the Resumable effect

### PR DESCRIPTION
**Pull request by j-mueller**
* Instead of discarding the continuations when get to a point where we don't have any more inputs, we store them in a map
* When running contract instances in the emulator we can hold on to the continuations and evaluate them when we receive the next event. This is what causes the speedup in the `plutus-use-cases` tests.
* When running compiled contracts in the SCB we still use the same method as before, serialising the state and re-running the computation upon every event
  * I'm starting to think maybe we shouldn't do that, and make the compiled contracts more stateful
    * We'd still store the state after every event, but only replay it when the process has been restarted. 

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [x] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by j-mueller** [comment]
this reduces the time to run the `plutus-use-cases` tests from 53s to 19s on my machine

File: plutus-contract/src/Language/Plutus/Contract/Types.hs [hunk 1]
```diff
@@ -209,32 +226,121 @@ handleError ::
...
+  Eff (ContractEffs s e) a -- ^ The contract
+  -> SuspendedContract e (Event s) (Handlers s) a
+suspend action =
+  mkResult
+    $ runSuspContractEffects
+      (0 :: CheckpointKey)
+      (mempty @CheckpointStore)
+      (handleContractEffs @_ @_ @(SuspendedContractEffects e (Event s)) action)
+
+runStep ::
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Types.hs:330]: haddock

**Review by michaelpj — COMMENTED** [review]
I got a bit fatigued trying to read this: I think it could do with some more doc comments to help the reader avoid getting lost!

**Comment by j-mueller** [comment]
Ok @michaelpj I added some comments and a note

**Review by michaelpj — APPROVED** [review]

**Pull request merged**